✨ Bring your own (encryption) key (BYOK)

akutz opened the pull request:
<!--
Thanks for sending a pull request!

Please add one of the following icons to the title of this PR:

    ⚠️ (:warning:, a major or breaking change)
    ✨ (:sparkles:, feature additions)
    🐛 (:bug:, patch and bugfixes)
    📖 (:book:, documentation or proposals)
    🌱 (:seedling:, minor or other)

Some other tips:

    1. If this is your first time filing a PR, please read our contributor
       guidelines for submitting a change at https://vm-operator.readthedocs.io/en/stable/start/contrib/submit-change/.
    2. If this PR is unfinished, please prefix the subject with "WIP:".

Finally, before filing the PR, please delete all of the HTML comments.
-->

**What does this PR do, and why is it needed?**

This patch adds support for bringing your own encryption key used to encrypt/recrypt VMs.



**Which issue(s) is/are addressed by this PR?** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes `NA`


**Are there any special notes for your reviewer**:

This PR will be rebased once #670 and #672 are merged.

<!-- Anything else you would like the reviewers to know about this PR. -->


**Please add a release note if necessary**:

<!--
Write your release note:

    1. Enter your extended release note in the below block. If the PR requires
       additional action from users switching to the new release, please include
       the string "action required".
    2. If a release note is not required, please write "NONE".
-->

```release-note
Support bringing your own (encryption) key (BYOK) to encrypt VM workloads
```

<!-- readthedocs-preview vm-operator start -->
----
📚 Documentation preview 📚: https://vm-operator--673.org.readthedocs.build/en/673/

<!-- readthedocs-preview vm-operator end -->

bryanv commented on pkg/providers/vsphere/vmprovider_vm.go:
```diff
@@ -1084,6 +1093,55 @@ func (vs *vSphereVMProvider) vmCreateGenConfigSpec(
+		return nil
+	}
+
+	isEncrypted := vmCtx.MoVM.Summary.Runtime.CryptoState != ""
```
I think this result in a Reconfigure() on every reconcile

rtsanev-github commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -420,6 +420,27 @@ type VirtualMachineSpec struct {
+	// used to encrypt this VM.
+	//
+	// Please note, this field is not required to encrypt the VM. If the
+	// underlying platform implements the vSphere Native Key Provider (NKP) then
```
if a default Key Provider is defined in the underlying platform...

It does not necessarily has to be native key provider, it can be standard key provider but it must be set as default in vCenter.

rtsanev-github commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -420,6 +420,27 @@ type VirtualMachineSpec struct {
+	// the VM may still be fully or partially encrypted depending on the
+	// specified storage and VM classes.
+	//
+	// If an NKP is present and an encryption storage class is selected, the
```
NKP > Default Key Provider 

here and below

rtsanev-github commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -420,6 +420,27 @@ type VirtualMachineSpec struct {
+	// specified storage and VM classes.
+	//
+	// If an NKP is present and an encryption storage class is selected, the
+	// VM's home files and non-PVC disks will be encrypted.
```
Maybe its worth noting somewhere that in case there is a default key provider in vCenter AND EncryptionClassName (ECN) is specified, then ECN takes precedence.

dougm commented on test/builder/vcsim_test_context.go:
```diff
@@ -648,18 +724,56 @@ func (c *TestContextForVCSim) setupK8sConfig(config VCSimTestConfig) {
+		})).To(Succeed())
+
+		c.EncryptedStorageClassName = "vm-encryption-policy"
+		c.EncryptedStorageProfileID = "4d5f673c-536f-11e6-beb8-9e71128cae77" // from vcsim
```
Use the exported const `pbm/simulator.DefaultEncryptionProfileID` instead ?
We also could/should add one for ID used in `StorageProfileID` above.

akutz replied:
Yep. It did not exist when I wrote this. I'll switch to that :)

dougm reviewed (approved):
lgtm, just a few minor comments/suggestions

akutz commented on pkg/providers/vsphere/session/session_vm_update.go:
```diff
@@ -1190,85 +1200,120 @@ func (s *Session) UpdateVirtualMachine(
+	}
+
+	if updateErr != nil {
+		updateErr = fmt.Errorf("updating state failed w %w", updateErr)
```
Ha, yeah, you and @bryanv literally had the exact same thought process :)

akutz commented on controllers/storageclass/storageclass_controller.go:
```diff
@@ -0,0 +1,157 @@
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	logger := ctrl.Log.WithName("StorageClass").WithValues("name", req.Name)
```
Yes, because here I'd rather not use reflection every time an object is reconciled. If I need to rebase this again, I'll add the reflected name to the reconciler struct and re-use it here.

aruneshpa commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -357,6 +361,65 @@ type VirtualMachineCdromSpec struct {
+	// If the underlying vSphere platform does not have a default key provider,
+	// then this field is required when specifying an encryption storage class
+	// and/or a VM Class with a vTPM.
+	ClassName string `json:"className,omitempty"`
```
What are your thoughts on calling this `encryptionClassName`?  There's no stuttering since the parent field is called `spec.crypto`.   IMHO, `spec.cryto.encryptionClassName` is more accurate than `spec.crypto.className`.

Not to mention we have another field called `className` which might be confusing.

akutz replied:
Let me think about it. I am leaning towards agreeing with you.

akutz replied:
Done, thanks @aruneshpa !

aruneshpa commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -357,6 +361,65 @@ type VirtualMachineCdromSpec struct {
+	// underlying platform has a default key provider, the VM may still be fully
+	// or partially encrypted depending on the specified storage and VM classes.
+	//
+	// If there is a default key provider and an encryption storage class is
```
Hmm.  I think this has to be more nuanced than that.

I think by "VM's home files", you mean files in VM_HOME directory.  But that also include PVC disks (FCDs can be created in either the designated fcd/ directory, or VM_HOME).

akutz replied:
No, by VM home files I mean any files in the VM directory that are *not* disks.

dilyar85 commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -357,6 +361,65 @@ type VirtualMachineCdromSpec struct {
+	// If there is a default key provider and an encryption storage class is
+	// selected, the VM's configuration and non-PVC disks will be encrypted.
+	//
+	// If there is a default key provider and a and a VM Class with a virtual,
```
Nit: extra "and a" here.

dilyar85 commented on controllers/storageclass/storageclass_controller.go:
```diff
@@ -0,0 +1,110 @@
+	}
+}
+
+// Reconciler reconciles a VirtualMachineClass object.
```
"VirtualMachineClass": should this be "StorageClass" based on `controlledType = &storagev1.StorageClass{}`  above?

dilyar85 commented on controllers/storageclass/storageclass_controller.go:
```diff
@@ -0,0 +1,110 @@
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	logger := ctrl.Log.WithName("StorageClass").WithValues("name", req.Name)
+	ctx = logr.NewContext(ctx, logger)
+
+	if !obj.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, r.ReconcileNormal(ctx, &obj)
```
Nit: move the `logger/ctx` update after the "if" statement since they aren't used until then?

dilyar85 commented on controllers/virtualmachine/virtualmachine/virtualmachine_controller.go:
```diff
@@ -167,6 +178,54 @@ func classToVMMapperFn(
+			}
+		}
+
+		logger.Info(
```
Also use `logger.V(4)` here to avoid log spam (since this will be called regardless of any requests exist)?

dilyar85 commented on docs/concepts/workloads/vm.md:
````diff
@@ -467,3 +467,118 @@ rockylinux_64Guest          Rocky Linux (64-bit)
+    message: Must be powered off to encrypt VM.
+```
+
+If the VM requested encryption and it was successful, then then condition will be:
````
extra "then"

akutz replied:
No, the extra then is a typo for `the` :)

dilyar85 commented on docs/concepts/workloads/vm.md:
```diff
@@ -467,3 +467,118 @@ rockylinux_64Guest          Rocky Linux (64-bit)
+| `status.crypto.providerID` | The provider ID used to encrypt the VM. |
+| `status.crypto.keyID` | The key ID used to encrypt the VM. |
+
+For example, the following is an example of the status of a VM encrypted with an encrypted storage class:
```
I noticed you have a mixed use of "encryption storage class" and "encrypted storage class" here and also in the code. Should we stick to one term ("encryption storage class") to keep it consistent with the API field?

dilyar85 commented on webhooks/virtualmachine/validation/virtualmachine_validator.go:
```diff
@@ -472,6 +476,42 @@ func (v validator) validateStorageClass(ctx *pkgctx.WebhookRequestContext, vm *v
+		encClassName = vm.Spec.Crypto.EncryptionClassName
+	)
+
+	if encClassName == "" {
+		return nil
+	}
+
+	if !pkgcfg.FromContext(ctx).Features.BringYourOwnEncryptionKey {
+
+		allErrs = append(allErrs, field.Invalid(
+			path, encClassName, "feature not enabled"))
```
This would allow users to specify `spec.crypto` fields (other than `encryptionClassName`) when the FSS is not enabled. Consider returning early if the `spec.crypto` is empty as well?

akutz replied:
Yes, because `crypto.useDefaultKeyProvider` is currently the default behavior for today. But yes, I should disallow setting that field to a non-nil value as well.

dilyar85 commented on webhooks/virtualmachine/validation/virtualmachine_validator.go:
```diff
@@ -472,6 +476,42 @@ func (v validator) validateStorageClass(ctx *pkgctx.WebhookRequestContext, vm *v
+
+	var (
+		allErrs      field.ErrorList
+		path         = field.NewPath("spec", "crypto", "className")
```
Should it be `encryptionClassName` at the end? This is not captured in test since you hard-coded the field path with this incorrect name.

dilyar85 reviewed (commented):
Dropping the first round of review for the API, controller and webhook code. Will post another review for the rest later.

aruneshpa commented on pkg/util/kube/storage.go:
```diff
@@ -0,0 +1,154 @@
+		ctrlclient.ObjectKey{Name: storageClassName},
+		&storageClass); err != nil {
+
+		if apierrors.IsNotFound(err) {
```
Here and below, simplify with:

`return false, client.IgnoreNotFound(err)`

akutz commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -357,6 +361,65 @@ type VirtualMachineCdromSpec struct {
+	// - Deploying a VirtualMachine with an encryption storage policy or vTPM
+	//   will fail.
+	//
+	// - If a VirtualMachine is encrypted with a provider other than the default
```
Hmm, fair. However, I wrote the API docs from the perspective of the reconciler, not the webhooks (create vs update). Let me think on this.

akutz replied:
So we cannot avoid this situation. If the underlying key provider is gone, then a user has to be able to set `spec.crypto.encryptionClassName` to empty to bypass the failure state of being unable to find the `EncryptionClass`. When they set it to empty, the VM is either rekeyed using the default provider or not, based on the value of `spec.crypto.useDefaultKeyProvider`. This is aligned with the existing behavior/options of VCD.

akutz commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -357,6 +361,65 @@ type VirtualMachineCdromSpec struct {
+	// - If a VirtualMachine is encrypted with a provider other than the default
+	//   key provider, the VM will be not be rekeyed.
+	//
+	//   Please note, this could result in a VirtualMachine that cannot be
```
Because if the underlying key or provider is removed, and the VM is not rekeyed to use the default provider, then you cannot power on the VM since it is encrypted and there is no key to decrypt it.

akutz replied:
I updated the comment to make it more clear.

akutz commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -694,6 +762,35 @@ type VirtualMachineAdvancedSpec struct {
+
+const (
+	VirtualMachineEncryptionTypeConfig VirtualMachineEncryptionType = "Config"
+	VirtualMachineEncryptionTypeDisks  VirtualMachineEncryptionType = "Disks"
```
Hmmm. True. I will likely update the status to indicate that `Disks` does not include PVCs.

akutz commented on api/v1alpha3/virtualmachine_types.go:
```diff
@@ -694,6 +762,35 @@ type VirtualMachineAdvancedSpec struct {
+type VirtualMachineEncryptionType string
+
+const (
+	VirtualMachineEncryptionTypeConfig VirtualMachineEncryptionType = "Config"
```
It does. But the documentation explains this (see `vm.md`). VMFiles is not descriptive either unless you already know.

akutz replied:
I updated the GoDocs/OpenAPI docs as well to explain this further.

akutz commented on pkg/util/kube/storage.go:
```diff
@@ -0,0 +1,154 @@
+	ctx context.Context,
+	k8sClient ctrlclient.Client,
+	storageClass storagev1.StorageClass,
+	encrypted bool) error {
```
Because nothing stops someone from modifying the underlying Storage Policy to no longer support encryption. This function is called by the StorageClass controller that ensures we always represent the current state of the underlying policy.

akutz commented on pkg/util/kube/storage.go:
```diff
@@ -0,0 +1,154 @@
+		//
+
+		if !encrypted {
+			// If the goal is to mark the StorageClass ss not encrypted, then we
```
I'm keeping it, and ack.

akutz commented on pkg/util/kube/storage.go:
```diff
@@ -0,0 +1,154 @@
+		// The ConfigMap does not already exist and the goal is to mark the
+		// StorageClass as encrypted, so go ahead and create the ConfigMap and
+		// return early.
+		obj.OwnerReferences = []metav1.OwnerReference{ownerRef}
```
I am using OwnerRefs because Sandeep will not allow me to annotate the StorageClasses and he did not like me placing a finalizer on them either when I was directly modifying the ConfigMap with the SC names. Neither the GoDocs nor the Kube docs mention any limit.

akutz commented on pkg/util/kube/storage.go:
```diff
@@ -0,0 +1,154 @@
+	}
+
+	// Create the patch used to update the ConfigMap.
+	objPatch := ctrlclient.StrategicMergeFrom(obj.DeepCopyObject().(ctrlclient.Object))
```
I wrote CreateAndPatch, so I am aware of what it does. The model I am using here is more efficient since I do not need to ever update *or* create the object if `encryption` is false.

dilyar85 reviewed (approved):
LGTM with some minor comments left above. Thanks @akutz for the clear docs and comments on this feature!

akutz commented on controllers/storageclass/storageclass_controller.go:
```diff
@@ -0,0 +1,110 @@
+	return ctrl.Result{}, r.ReconcileNormal(ctx, &obj)
+}
+
+func (r *Reconciler) ReconcileNormal(
```
That is correct. Any errors will be reported via controller-runtime. I do not believe in happy path logging except where it provides value. Here it does nothing but clutter logs.

akutz commented on pkg/util/kube/storage.go:
```diff
@@ -0,0 +1,154 @@
+				Name:      internal.EncryptedStorageClassNamesConfigMapName,
+			},
+		}
+		ownerRef = internal.GetOwnerRefForStorageClass(storageClass)
```
Then you did not keep reading :) It is used for testing, hence the `internal` package. That way I can use it between prod and test without exposing it to beyond these two packages.

aruneshpa commented on pkg/vmconfig/vmconfig_reconciler.go:
```diff
@@ -0,0 +1,51 @@
+
+// Reconciler is a type that modifies a VirtualMachine's configuration before
+// and/or after creating or updating the underlying, platform resource.
+type Reconciler interface {
```
I know you are trying to introduce a pattern for this.  But reading through this code and the diff in the PR, my opinion is that we introduce this pattern when we have a _need_ for it.  Right now, it seems a bit too complicated for something that might not even be used by other components (if we never get to breaking the current pre/post power on reconfigure thing in this mode).

If you don't mind, I would recommend shelving all this, keeping this change specific to reconciling the StorageClasses and the VMs that are referencing them.

akutz replied:
I appreciate the feedback, but I am going to keep this. It is part of my vision going forward, and this change highlights *that* the pattern works and *how* the pattern works. I initially did this change without the pattern, and it ended up looking very similar, just without the wrapper. All of VM Op should start adopting this pattern in order to make our reconciler less fragile than it is and so difficult to change.

aruneshpa commented on pkg/vmconfig/crypto/crypto_reconciler_pre.go:
```diff
@@ -0,0 +1,690 @@
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha3"
+)
+
+type ptrCfgSpec = *vimtypes.VirtualMachineConfigSpec
```
Hmm.  Is this a pattern?  What does this buy us compared to just using `*vimtypes.VirtualMachineConfigSpec`?

akutz replied:
It is used elsewhere in our codebase, yes. The types are exactly the same, and it enables more concise code. Even intellisense shows what it actually is.

aruneshpa commented on pkg/vmconfig/crypto/crypto_reconciler.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func (r Reason) String() string {
+	return bitmask.String(r)
```
Andrew, I would rather have us just use the reasons explicitly, as opposed to define a type for them, and then use bitmasks to figure out the reasons that apply.  This will be hard to maintain and I don't see many usecases for this pattern. 

My recommendation would be to shelve the bitmask manipulation entirely and references the reasons explicitly where needed.

akutz replied:
The reasons *are* used explicitly, but there can be many different kinds and combinations. So I disagree there is not a use case for this in the future. There is literally a use case for it in this PR. The documentation in `vm.md` clearly outlines the reasons to the user. I appreciate the feedback, but I am going to keep this as-is.

akutz replied:
FWIW, here is [a test case](https://github.com/vmware-tanzu/vm-operator/pull/673/files#diff-916fea486b6a687a518cffddb850c48ae19a10f9f42864ac3e3507126a6cc233R324-R341) that shows when the reason can be both `InvalidState` _and_ `InvalidChanges`

akutz replied:
All that said, this _is_ something I am willing to revisit. I will give it more thought, but I am not going to let it block this PR. After it merges, I'll sit down and consider how many possible combinations there can be in practical terms. If it is just `InvalidState` and `InvalidChanges`, then I can have a distinct reason for that. Let me look.

akutz commented:
@aruneshpa I have addressed all of your feedback on the PR. While I did not accept it all, and stated my reasons, unless you have a strong objection, I am going to merge this PR so it does not continue to grow. We can revisit your concerns post-merge if that is alright?

aruneshpa commented:
> @aruneshpa I have addressed all of your feedback on the PR. While I did not accept it all, and stated my reasons, unless you have a strong objection, I am going to merge this PR so it does not continue to grow. We can revisit your concerns post-merge if that is alright?

Sounds good to me.  Thanks!
